src/control/cmd/daos_server/storage.go

```diff
@@ -35,97 +35,184 @@ type storagePrepareCmd struct {
-		// Pass evaluated state to avoid running GetScmState() twice.
-		resp, err := cmd.scs.ScmPrepare(storage.ScmPrepareRequest{Reset: cmd.Reset})
-		if err != nil {
-			return common.ConcatErrors(scanErrors, err)
-		}
-		if resp.RebootRequired {
-			cmd.log.Info(storage.ScmMsgRebootRequired)
-		} else if len(resp.Namespaces) > 0 {
-			var bld strings.Builder
-			if err := pretty.PrintScmNamespaces(resp.Namespaces, &bld); err != nil {
-				return err
-			}
-			cmd.log.Infof("SCM namespaces:\n%s\n", bld.String())
-		} else {
-			cmd.log.Info("no SCM namespaces")
+		return nil
+	}
+
+	if state == storage.ScmStateUnknown {
+		return scanErrors.add(errors.New("failed to report state"))
+	}
+	if state == storage.ScmStateNoModules {
+		return scanErrors.add(storage.FaultScmNoModules)
+	}
+
+	if cmd.Reset {
+		// Respond to resultant state on prepare reset.
+		if state == storage.ScmStateNoRegions {
+			cmd.log.Info("SCM reset successfully!")
+			return nil
 		}
-	} else if prepScm {
-		cmd.log.Info("No SCM modules detected; skipping operation")
+
+		return scanErrors.add(errors.Errorf("unexpected state %q after prepare reset", state))
 	}
 
-	if len(scanErrors) > 0 {
-		return common.ConcatErrors(scanErrors, nil)
+	// Respond to state reported by prepare setup.
+	switch state {
+	case storage.ScmStateNoRegions:
+		scanErrors.add(errors.New("failed to create regions"))
+	case storage.ScmStateFreeCapacity:
+		scanErrors.add(errors.New("failed to create namespaces"))
+	case storage.ScmStateNoFreeCapacity:
+		if len(resp.Namespaces) == 0 {
+			scanErrors.add(errors.New("failed to find namespaces"))
+			break
+		}
+		// Namespaces exist.
+		var bld strings.Builder
+		if err := pretty.PrintScmNamespaces(resp.Namespaces, &bld); err != nil {
+			return err
+		}
+		cmd.log.Infof("%s\n", bld.String())
+	default:
+		scanErrors.add(errors.Errorf("unexpected state %q after prepare reset", state))
 	}
 
 	return nil
 }
 
+func (cmd *storagePrepareCmd) Execute(args []string) error {
+	if cmd.HelperLogFile != "" {
+		if err := os.Setenv(pbin.DaosAdminLogFileEnvVar, cmd.HelperLogFile); err != nil {
+			cmd.log.Errorf("unable to configure privileged helper logging: %s", err)
+		}
+	}
+
+	// This is a little ugly, but allows for easier unit testing.
+	// FIXME: With the benefit of hindsight, it seems apparent
+	// that we should have made these Execute() methods thin
+	// wrappers around more easily-testable functions.
+	if cmd.scs == nil {
+		cmd.scs = server.NewStorageControlService(cmd.log, config.DefaultServer().Engines)
+	}
+
+	scanErrors := make(errs, 0, 2)
+
+	if err := cmd.prepNvme(&scanErrors, cmd.scs.NvmePrepare); err != nil {
+		return err
+	}
+
+	if err := cmd.prepScm(&scanErrors, cmd.scs.ScmPrepare); err != nil {
+		return err
+	}
+
+	switch len(scanErrors) {
+	case 0:
+		return nil
+	case 1:
+		return scanErrors[0]
+	default:
+		// When calling ConcatErrors the Fault resolution is lost.
+		return common.ConcatErrors(scanErrors, nil)
+	}
```

Comment:
I think this would benefit greatly from some comments explaining why we are accumulating errors to display later. I could go study the code to try to understand the intent, but I would rather read a comment explaining the unusual pattern.

I am somewhat wary of not only reintroducing this pattern (it was removed on the server side startup scan, right?), but formalizing it with a new type.

To put it another way: Why shouldn't an error shut the entire operation down? It seems like we should be aiming to simplify the code in order to reduce the maintainable surface and opportunity for bugs and unexpected behavior.

Reply:
I was trying to restrict the scope of the PR to not separate the syntax into individual commands for scm/nvme, which will remove all this error concatenation code. That will be done for 2.4 in a follow-up PR and will tidy this away. Also please note that this is not being "reintroduced", it's just been moved.

Reply:
OK, thanks for the explanation. Given that master has now been re-opened for 2.4+ landings, do you think it's worthwhile to just go ahead and clean that stuff up now instead of landing in this intermediate state?

Reply:
I would prefer to land this now and address as a follow-on.

Reply:
https://daosio.atlassian.net/browse/DAOS-10284 will result in the removal of the error concatenation code